### PR TITLE
feat: make options having a single source of truth

### PR DESCRIPTION
**Pull request by hyf0**
## Summary
- use `zod` as options validator
- Make options having a single source of truth
- With this integration, the options type would be derived from the schema.

### <samp>🤖 Generated by Copilot at d3324a0</samp>

Added `zod` as a dependency and used it to create a new configuration schema for `rspack` in `packages/rspack/src/config/zod`. This schema validates and transforms the configuration options and replaces the old type system in `packages/rspack/src/config/types.ts`. The schema is composed of several modules that define the sub-properties of the configuration.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at d3324a0</samp>

*  Simplify and validate the configuration schema using `zod` ([link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-142a5c1a681bf6859bd9e53bbfe87fc9dd2d05e8350b0fb50bfcd652a9e36cbbL71-R72), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L1183-R1185), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L1283-L1285), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L1307-L1309), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L1315-L1320), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L1940-L2050), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L2121-L2182), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L2205-R1938), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L2261-L2283), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fR17-R18), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fL28-R30), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-bd8525628aa18fe2735262e8e0528ec0b40d0bac1d18250050ba0e2e112e2a50R1-R28), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-2aaf2d605b13329eee31f67946b79bd691b267f1a6274851df7a60276b1e9d37R1-R6), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-be4ee8204dd2e7caa919ec90f6ef3df2765462bea7ff5fb3be8efc462def855fR1-R33), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-4efd9db7546a820ca2c06f0f52e2942bb132a81e93f6aac326eaba190ebd2c88R1-R23), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-d0d329e7eba5c2bc8a9e487bf452b933834266020393c24d092acd67b8293ce9R1-R12), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-8c9c55e119547fb481071dee846c579a8df388a933cd5628c0249875644a8478R1-R9), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-ab1e5d9e877cb4e37eb81597cc9703010c4475c7dc8644087bfaa41cf23ffa95R1-R27), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-cb662a6075e4dfd420966ecf55a77482ea00ab31239733eaf16eee00f2168c1eR1-R11), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-9a73f0c83f0cf8363ded94e6e095ba53087a6ae86ef35d56df11f456a446a571R1-R50), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-3c08b09f5e0814c49ac42eaaa62527b4836fb8e52e8d6ce1f9307b710c2863f4R1-R16), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-bfc1f66f5bb4feef96b1ba248bcfdce90d237a46edcfa72ff530c3d0e193dc40R1-R22), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-a87fa326cc04dba4ec79610c6e4e1f001873aec11de2c2190c375470a7a5d9a8R1-R94), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-32a7f145fcd08de09af167524aa83cdfc06711788b32b25991a3a11675b6fc4aR1-R14), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-cefd3d63a68dce1408415d104eb1619aa13286c73c1c409dbece57e02d3517abR1-R34), [link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-e63a3b8814d8c67470d1e315c43b58f648e3e3b22a266598a23a5d2a44dbf2a7R1-R22))
* Cast the `ignore` parameter to a function type in `getNormalizedRspackOptions` ([link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-78648f723799beccc94c4b2e506997d12d1ace8216aeb31c0731c37a6dce024aL29-R29))
* Sort the dependencies in `package.json` alphabetically ([link](https://github.com/web-infra-dev/rspack/pull/3378/files?diff=unified&w=0#diff-142a5c1a681bf6859bd9e53bbfe87fc9dd2d05e8350b0fb50bfcd652a9e36cbbL61-R63))

</details>

### CODE REVIEW

File: packages/rspack/src/config/types.ts [hunk 1]
```diff
@@ -25,7 +27,7 @@ export type {
 
 export type Configuration = RspackOptions;
 
-export interface RspackOptions {
+export interface LegacyRspackOptions {
```
Comment (Boshen) [inline packages/rspack/src/config/types.ts:30]: Can you add a deprecation comment so collaborators know why it's called legacy?

File: packages/rspack/src/config/zod/builtins.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { z } from "zod";
+
+export function builtins() {
+	// TODO(hyf0): need to enable strict mode sometimes
```
Comment (Boshen) [inline packages/rspack/src/config/zod/builtins.ts:4]: Can you be more clear on when "sometimes" is? These kind of todo comments tend to stay for ever, and we don't know when to remove these.

File: packages/rspack/src/config/zod/output.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		.or(z.string());
+}
+
+function wasmLoadingType() {
+	return z.enum(["...", "fetch-streaming", "fetch", "async-node"]);
+}
+
+function libraryType() {
+	return z.enum([
+		"...",
```
Comment (Boshen) [inline packages/rspack/src/config/zod/output.ts:15]: What's this?

Comment (hyf0) [reply packages/rspack/src/config/zod/output.ts:15]: https://github.com/web-infra-dev/rspack/blob/f20a34d3c9892ef1e2070f208c932dfd7c3bdfe2/packages/rspack/src/config/normalization.ts#L102-L103

File: packages/rspack/src/config/types.ts (hunk 1 above)
Comment (hyf0) [reply packages/rspack/src/config/types.ts]: I removed this type.

**Comment by ahabhgk** [comment]
Is there any performance compare about options validation

**Comment by hyf0** [comment]
> Is there any performance compare about options validation, precompiled ajv vs zod

Nope. But if there are performance problems, we could fallback to json-shchema. 

`zod` could be compiled to json-schema.

**Review by ahabhgk — APPROVED** [review]

**Pull request merged**